testnet/stacks-node/src/nakamoto_node/signer_coordinator.rs

```diff
@@ -0,0 +1,436 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::BTreeMap;
+use std::sync::atomic::AtomicBool;
+use std::sync::{Arc, Condvar, Mutex};
+use std::thread::JoinHandle;
+
+use hashbrown::{HashMap, HashSet};
+use libsigner::v0::messages::{MinerSlotID, SignerMessage as SignerMessageV0};
+use libsigner::{BlockProposal, SignerSession, StackerDBSession};
+use stacks::burnchains::Burnchain;
+use stacks::chainstate::burn::db::sortdb::SortitionDB;
+use stacks::chainstate::burn::{BlockSnapshot, ConsensusHash};
+use stacks::chainstate::nakamoto::{NakamotoBlock, NakamotoChainState};
+use stacks::chainstate::stacks::boot::{RewardSet, MINERS_NAME};
+use stacks::chainstate::stacks::db::StacksChainState;
+use stacks::chainstate::stacks::Error as ChainstateError;
+use stacks::codec::StacksMessageCodec;
+use stacks::libstackerdb::StackerDBChunkData;
+use stacks::net::stackerdb::StackerDBs;
+use stacks::types::chainstate::{StacksBlockId, StacksPrivateKey, StacksPublicKey};
+use stacks::util::hash::Sha512Trunc256Sum;
+use stacks::util::secp256k1::MessageSignature;
+use stacks::util_lib::boot::boot_code_id;
+
+use super::Error as NakamotoNodeError;
+use crate::event_dispatcher::StackerDBChannel;
+use crate::nakamoto_node::stackerdb_listener::{
+    BlockStatus, StackerDBListener, TimestampInfo, EVENT_RECEIVER_POLL,
+};
+use crate::neon::Counters;
+use crate::Config;
+
+/// Helper function to determine if signer threshold has been reached for a block
+fn is_threshold_reached(
+    status: Option<&BlockStatus>,
+    weight_threshold: u32,
+    total_weight: u32,
+) -> bool {
+    match status {
+        Some(status) => {
+            status.total_weight_signed >= weight_threshold
+                || status.total_reject_weight.saturating_add(weight_threshold) > total_weight
+        }
+        None => true,
+    }
+}
+
+/// The state of the signer database listener, used by the miner thread to
+/// interact with the signer listener.
+pub struct SignerCoordinator {
+    /// The private key used to sign messages from the miner
+    message_key: StacksPrivateKey,
+    /// Is this mainnet?
+    is_mainnet: bool,
+    /// The session for writing to the miners contract in the stackerdb
+    miners_session: StackerDBSession,
+    /// The total weight of all signers
+    total_weight: u32,
+    /// The weight threshold for block approval
+    weight_threshold: u32,
+    /// Tracks signatures for blocks
+    ///   - key: Sha512Trunc256Sum (signer signature hash)
+    ///   - value: BlockStatus
+    blocks: Arc<(Mutex<HashMap<Sha512Trunc256Sum, BlockStatus>>, Condvar)>,
```

Comment:
It seems like you have a low-lift opportunity for a future-proofed refactoring here.

In other places, if struct `Foo` needs to share its state with other threads, it has a corresponding `FooComms` struct which contains all of these `Arc<Mutex<..>>`-wrapped data.  The `FooComms` struct then exposes getters and setters for the wrapped data, so the caller doesn't need to bother with e.g. locking the mutex or dereferencing the `Arc` the right way.  The `Foo` implementation has a method or instantiating a `FooComms`.

It seems like you could have a `StackerDBSessionComms` struct which contained `blocks` and `signer_idle_timestamps`, and gets instantiated from the `StackerDBSession`.

Reply:
This is done with `StackerDBListenerComms`.